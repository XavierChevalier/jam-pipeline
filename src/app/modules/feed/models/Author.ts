import { InferType, object, string } from 'yup'

export const Author = object({
  name: string().required(),
  avatar: string().required().url(),
})

export type Author = InferType<typeof Author>
