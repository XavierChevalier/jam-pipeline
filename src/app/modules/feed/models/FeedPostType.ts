export enum AvailableFeedPostType {
  userShareSong = 'userShareSong',
  userShareAlbum = 'userShareAlbum',
  userSharePlaylist = 'userSharePlaylist',
  artistReleaseSong = 'artistReleaseSong',
  artistReleaseAlbum = 'artistReleaseAlbum',
}
