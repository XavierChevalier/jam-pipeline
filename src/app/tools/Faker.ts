export const generateImageUrl = (size = 50) => `https://picsum.photos/${size}`
