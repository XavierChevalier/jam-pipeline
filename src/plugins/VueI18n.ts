import { createI18n } from 'vue-i18n'
import { Ref } from 'vue'

export const i18n = createI18n({
  legacy: false,
  globalInjection: true,
  locale: 'en',
  fallbackLocale: 'en',
})

export function changeLocale(locale: 'en' | 'fr') {
  if (i18n.mode === 'legacy') {
    i18n.global.locale = locale
    return
  }

  ;(i18n.global.locale as unknown as Ref<string>).value = locale
}
