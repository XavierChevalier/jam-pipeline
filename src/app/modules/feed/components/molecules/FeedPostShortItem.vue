<script setup lang="ts">
  import AppAvatar from '@/app/components/atoms/AppAvatar.vue'
  import { localeFormatDistance } from '@/app/tools/date/Date'
  import { computed, PropType } from 'vue'
  import { Author } from '@/app/modules/feed/models/Author'
  import { isPropertyValid } from '@/app/tools/component-properties/PropertyValidator'
  import { AvailableFeedPostType } from '@/app/modules/feed/models/FeedPostType'
  import { useI18n } from 'vue-i18n'

  const { t } = useI18n()

  const props = defineProps({
    author: {
      type: Object as PropType<Author>,
      required: true,
      validator: isPropertyValid(Author),
    },
    postType: {
      type: String as PropType<keyof typeof AvailableFeedPostType>,
      required: true,
      validator: (value: string) =>
        Object.keys(AvailableFeedPostType).includes(value),
    },
    publicationDate: {
      type: Date,
      required: true,
    },
  })
  const publicationDateDistance = computed(() =>
    localeFormatDistance(props.publicationDate, new Date(), {
      addSuffix: false,
    })
  )
</script>

<template>
  <div class="flex gap-3 items-center">
    <AppAvatar size="sm" :user-name="author.name" :url="author.avatar" />
    <span>{{ t(`postType.${postType}`, { authorName: author.name }) }}</span>
    <div class="flex relative flex-shrink-0 gap-2 items-center ml-auto">
      <div
        class="absolute -left-3 w-0 h-0 rounded-full border-3 border-primary"
      />
      <span class="text-neutral">{{ publicationDateDistance }}</span>
    </div>
  </div>
</template>

<i18n lang="yaml">
en:
  postType:
    userShareSong: '{authorName} shared a song'
    userShareAlbum: '{authorName} shared an album'
    userSharePlaylist: '{authorName} shared a playlist'
    artistReleaseSong: '{authorName} released a new song'
    artistReleaseAlbum: '{authorName} released a new album'
fr:
  postType:
    userShareSong: '{authorName} a partagé un morceau'
    userShareAlbum: '{authorName} a partagé un album'
    userSharePlaylist: '{authorName} a partagé une playlist'
    artistReleaseSong: '{authorName} a sortie un nouveau morceau'
    artistReleaseAlbum: '{authorName} a sortie un nouvel album'
</i18n>
