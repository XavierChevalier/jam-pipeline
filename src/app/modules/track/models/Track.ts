import { TrackCover } from '@/app/modules/track/models/TrackCover'
import { array, InferType, number, object, string } from 'yup'

export const TrackTitle = string().required()
export const TrackBands = array(string().required()).required().min(1)

export const Track = object({
  cover: TrackCover,
  title: TrackTitle,
  bands: TrackBands,
  durationMs: number().positive().required(),
})
export type Track = InferType<typeof Track>
