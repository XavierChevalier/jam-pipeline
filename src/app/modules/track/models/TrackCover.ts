import { TrackBands, TrackTitle } from '@/app/modules/track/models/Track'
import { InferType, object, string } from 'yup'

export const TrackCover = string().url().required()
export type TrackCover = InferType<typeof TrackCover>

export const TrackOfTrackCover = object({
  cover: TrackCover,
  title: TrackTitle,
  bands: TrackBands,
})
export type TrackOfTrackCover = InferType<typeof TrackOfTrackCover>
