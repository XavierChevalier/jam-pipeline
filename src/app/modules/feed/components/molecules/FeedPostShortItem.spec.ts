import FeedPostShortItem from './FeedPostShortItem.vue'
import * as Stories from '@/app/modules/feed/components/molecules/FeedPostShortItem.stories'
import { Author } from '@/app/modules/feed/models/Author'
import { omit } from 'lodash-es'
import { AvailableFeedPostType } from '@/app/modules/feed/models/FeedPostType'
import { changeLocale } from '@/plugins/VueI18n'
import { shallowMountWithPlugins } from '@/tests/tools/MountWithPlugins'
import { StorybookTestsGenerator } from '@/tests/tests-generators/StorybookTestsGenerator'
import { generateImageUrl } from '@/app/tools/Faker'

describe('FeedPostShortItem', () => {
  StorybookTestsGenerator.fromStoriesExports(Stories)
    .snapshotEachStories()
    .property<Author>('author', (property) => {
      const author = { name: 'John Gomm', avatar: generateImageUrl() }
      property
        .itShouldBeDefined()
        .itShouldBeRequired()
        .itShouldHaveAValidatorFunction()
        .itShouldFailWith('missing name', omit(author, ['name']))
        .itShouldFailWith('a null avatar', { name: 'Test', avatar: null })
        .itShouldFailWith('a null name and avatar', {
          name: null,
          avatar: null,
        })
    })
    .property<keyof typeof AvailableFeedPostType>('postType', (property) => {
      property
        .itShouldBeDefined()
        .itShouldBeTypeOf(String)
        .itShouldBeRequired()
        .itShouldHaveAValidatorFunction()
        .itShouldSuccessWith('userShareSong')
        .itShouldSuccessWith('userShareAlbum')
        .itShouldSuccessWith('userSharePlaylist')
        .itShouldSuccessWith('artistReleaseSong')
        .itShouldSuccessWith('artistReleaseAlbum')
    })
    .property<Date>('publicationDate', (property) => {
      property.itShouldBeDefined().itShouldBeTypeOf(Date).itShouldBeRequired()
    })

  it.each`
    postType                                    | lang    | expectedDisplayText
    ${AvailableFeedPostType.userShareSong}      | ${'fr'} | ${'John Gomm a partagé un morceau'}
    ${AvailableFeedPostType.userShareSong}      | ${'en'} | ${'John Gomm shared a song'}
    ${AvailableFeedPostType.userShareAlbum}     | ${'fr'} | ${'John Gomm a partagé un album'}
    ${AvailableFeedPostType.userShareAlbum}     | ${'en'} | ${'John Gomm shared an album'}
    ${AvailableFeedPostType.userSharePlaylist}  | ${'fr'} | ${'John Gomm a partagé une playlist'}
    ${AvailableFeedPostType.userSharePlaylist}  | ${'en'} | ${'John Gomm shared a playlist'}
    ${AvailableFeedPostType.artistReleaseSong}  | ${'fr'} | ${'John Gomm a sortie un nouveau morceau'}
    ${AvailableFeedPostType.artistReleaseSong}  | ${'en'} | ${'John Gomm released a new song'}
    ${AvailableFeedPostType.artistReleaseAlbum} | ${'fr'} | ${'John Gomm a sortie un nouvel album'}
    ${AvailableFeedPostType.artistReleaseAlbum} | ${'en'} | ${'John Gomm released a new album'}
  `(
    'should display "$expectedDisplayText" on "$postType" postType',
    ({ postType, lang, expectedDisplayText }) => {
      changeLocale(lang)
      const wrapper = shallowMountWithPlugins(FeedPostShortItem, {
        props: {
          author: { name: 'John Gomm', avatar: generateImageUrl() },
          publicationDate: new Date(),
          postType,
        },
      })

      expect(wrapper.text()).toContain(expectedDisplayText)
    }
  )
})
