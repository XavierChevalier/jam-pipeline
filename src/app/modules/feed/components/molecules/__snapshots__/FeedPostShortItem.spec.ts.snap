// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedPostShortItem stories snapshots [snapshot] should render the "ArtistReleaseAlbum" component story correctly 1`] = `
<div
  class="flex gap-3 items-center"
>
  <anonymous-stub
    size="sm"
    url="https://picsum.photos/50"
    username="TheBaronless"
  />
  <span>
    TheBaronless released a new album
  </span>
  <div
    class="flex relative flex-shrink-0 gap-2 items-center ml-auto"
  >
    <div
      class="absolute -left-3 w-0 h-0 rounded-full border-3 border-primary"
    />
    <span
      class="text-neutral"
    >
      less than a minute
    </span>
  </div>
</div>
`;

exports[`FeedPostShortItem stories snapshots [snapshot] should render the "ArtistReleaseSong" component story correctly 1`] = `
<div
  class="flex gap-3 items-center"
>
  <anonymous-stub
    size="sm"
    url="https://picsum.photos/50"
    username="TheBaronless"
  />
  <span>
    TheBaronless released a new song
  </span>
  <div
    class="flex relative flex-shrink-0 gap-2 items-center ml-auto"
  >
    <div
      class="absolute -left-3 w-0 h-0 rounded-full border-3 border-primary"
    />
    <span
      class="text-neutral"
    >
      less than a minute
    </span>
  </div>
</div>
`;

exports[`FeedPostShortItem stories snapshots [snapshot] should render the "UserShareAlbum" component story correctly 1`] = `
<div
  class="flex gap-3 items-center"
>
  <anonymous-stub
    size="sm"
    url="https://picsum.photos/50"
    username="TheBaronless"
  />
  <span>
    TheBaronless shared an album
  </span>
  <div
    class="flex relative flex-shrink-0 gap-2 items-center ml-auto"
  >
    <div
      class="absolute -left-3 w-0 h-0 rounded-full border-3 border-primary"
    />
    <span
      class="text-neutral"
    >
      less than a minute
    </span>
  </div>
</div>
`;

exports[`FeedPostShortItem stories snapshots [snapshot] should render the "UserSharePlaylist" component story correctly 1`] = `
<div
  class="flex gap-3 items-center"
>
  <anonymous-stub
    size="sm"
    url="https://picsum.photos/50"
    username="TheBaronless"
  />
  <span>
    TheBaronless shared a playlist
  </span>
  <div
    class="flex relative flex-shrink-0 gap-2 items-center ml-auto"
  >
    <div
      class="absolute -left-3 w-0 h-0 rounded-full border-3 border-primary"
    />
    <span
      class="text-neutral"
    >
      less than a minute
    </span>
  </div>
</div>
`;

exports[`FeedPostShortItem stories snapshots [snapshot] should render the "UserShareSong" component story correctly 1`] = `
<div
  class="flex gap-3 items-center"
>
  <anonymous-stub
    size="sm"
    url="https://picsum.photos/50"
    username="TheBaronless"
  />
  <span>
    TheBaronless shared a song
  </span>
  <div
    class="flex relative flex-shrink-0 gap-2 items-center ml-auto"
  >
    <div
      class="absolute -left-3 w-0 h-0 rounded-full border-3 border-primary"
    />
    <span
      class="text-neutral"
    >
      less than a minute
    </span>
  </div>
</div>
`;
